Sources/Datasets/datasets/image/MNIST.swift

```diff
@@ -1,99 +1,108 @@
-      volume={2},
-      year={2010}
-    }
-    """)
-
-public struct MNISTIterator: S5TFDatasetIterator {
-    typealias _Dataset = MNIST // swiftlint:disable:this type_name
-    private var dataset: MNIST
+    homepage: URL(string: "http://yann.lecun.com/exdb/mnist/")!
+)
 
+public struct MNISTDataLoader: S5TFDataLoader {
     private var index = 0
 
-    public init(dataset: MNIST) {
-        self.dataset = dataset
-    }
-
-    mutating public func next() -> MNISTBatch? {
-        guard index < dataset.data.count else { return nil }
-        let data = dataset.data[index]
-        index += 1
-        return MNISTBatch(data: data[0], labels: data[1])
-    }
-}
+    let split: S5TFSplit
+    let batchSize: Int
+    let data: Tensor<UInt8>
+    let labels: Tensor<Int64>
 
-public struct MNIST: S5TFDataset {
-    typealias _Iterator = MNISTIterator // swiftlint:disable:this type_name
-
-    internal var data = [[1, 2], [3, 4], [5, 6]]
-
-    public func batched(_ batchSize: Int) -> MNISTIterator {
-        return MNISTIterator(dataset: self)
-    }
-
-    public var train: MNIST {
-        guard self.split == .undefined else { fatalError("This property can only be accessed from an undefined split.") }
-        return MNIST(split: .train)
-    }
-    public var validation: MNIST {
-        fatalError("The split does not exist for this dataset.")
-    }
-    public var test: MNIST {
-        guard self.split == .undefined else { fatalError("This property can only be accessed from an undefined split.") }
-        return MNIST(split: .test)
-    }
-    public var all: MNIST {
-        guard self.split == .undefined else { fatalError("This property can only be accessed from an undefined split.") }
-        return MNIST(split: .all)
-    }
-
-    public let split: S5TFSplit
-
-    public init() { self.split = .undefined }
-    private init(split: S5TFSplit) {
+    internal init(split: S5TFSplit, batchSize: Int) {
+        self.batchSize = batchSize
         self.split = split
-        let downloader = Downloader()
+
+        let semaphore = DispatchSemaphore(value: 0)
         let baseURL = "https://storage.googleapis.com/cvdf-datasets/mnist"
-        func getTrain() {
-            downloader.download(fileAt: URL(string: baseURL + "/train-images-idx3-ubyte.gz")!,
-            					cacheName: "mnist", fileName: "mnist_train_images") { url, error in }
-            downloader.download(fileAt: URL(string: baseURL + "/train-labels-idx1-ubyte.gz")!,
-            					cacheName: "mnist", fileName: "mnist_train_labels") { url, error in }
-        }
-        func getTest() {
-            downloader.download(fileAt: URL(string: baseURL + "/t10k-images-idx3-ubyte.gz")!,
-            					cacheName: "mnist", fileName: "mnist_test_images") { url, error in }
-            downloader.download(fileAt: URL(string: baseURL + "/t10k-labels-idx1-ubyte.gz")!,
-            					cacheName: "mnist", fileName: "mnist_test_labels") { url, error in }
-        }
```

Comment:
Nice!

One small thing: I think baseURL should be of type URL, so `let baseURL = URL(string: "https://storage.googleapis.com/cvdf-datasets/mnist")!`. Then you can change this `URL(string: baseURL + "/train-images-idx3-ubyte.gz")!` to this `baseURL.appendingPathComponent("train-images-idx3-ubyte.gz")`.